vortex-file/src/read/coalescer.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::VecDeque;
+use std::io;
+use std::io::ErrorKind;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{ready, Context, Poll, Waker};
+
+use futures::Stream;
+use futures_util::future::BoxFuture;
+use futures_util::{FutureExt, StreamExt};
+use vortex_array::ArrayData;
+use vortex_error::{vortex_err, vortex_panic, VortexExpect, VortexResult};
+use vortex_io::{Dispatch, IoDispatcher, VortexReadAt, VortexReadRanges};
+
+use crate::{LayoutMessageCache, LayoutReader, Message, MessageLocator, MessageRead, RowMask};
+
+const NUM_TO_COALESCE: usize = 8;
+
+pub trait RowMaskReader<V> {
+    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<V>>>;
+}
+
+pub struct MaskLayoutReader {
+    layout: Box<dyn LayoutReader>,
+}
+
+impl MaskLayoutReader {
+    pub fn new(layout: Box<dyn LayoutReader>) -> Self {
+        Self { layout }
+    }
+}
+
+impl RowMaskReader<ArrayData> for MaskLayoutReader {
+    /// Read given mask out of the reader
+    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<ArrayData>>> {
+        self.layout.read_selection(mask)
+    }
+}
+
+enum RowMaskState<V> {
+    Pending(RowMask),
+    Ready(V),
+}
+
+pub struct Coalescer<R, S, V, RM> {
+    values_iter: S,
+    in_flight: Option<BoxFuture<'static, io::Result<Vec<Message>>>>,
+    queued: VecDeque<RowMaskState<V>>,
+    io_read: VortexReadRanges<R>,
+    dispatcher: Arc<IoDispatcher>,
+    row_mask_reader: RM,
+    cache: Arc<RwLock<LayoutMessageCache>>,
+}
+
+impl<R, S, V, RM> Coalescer<R, S, V, RM>
+where
+    R: VortexReadAt,
+    S: Stream<Item = VortexResult<RowMask>> + Unpin,
+    RM: RowMaskReader<V>,
+{
+    pub fn new(
+        read: R,
+        dispatcher: Arc<IoDispatcher>,
+        values_iter: S,
+        row_mask_reader: RM,
+        cache: Arc<RwLock<LayoutMessageCache>>,
+    ) -> Self {
+        Self {
+            values_iter,
+            in_flight: None,
+            queued: VecDeque::new(),
+            io_read: VortexReadRanges::new(read, dispatcher.clone(), 1 << 20),
```

Comment:
lol @ magic `1<<20` (presumably 1MiB?)

Reply:
yes, we should get rid of it and instead messure the value in the reader